Wait on readiness service on start

rjernst opened the pull request:
Node.start() currently does not return until the node is serving http. However, other parts of the system may not be quite ready. For example, file settings may not have been applied, and we may not yet have a master node. The readiness service's purpose is to identify when the node is actually ready for serving requests.

This commit adjusts the end of startup to wait on the readiness service being ready. The ramifications are that the main Elasticsearch thread will not exit until the node is actually ready to serve requests, and the cli will not exit (when in daemon mode) until the node is ready.

closes #136955

rjernst committed:
Wait on readiness service on start

Node.start() currently does not return until the node is serving http.
However, other parts of the system may not be quite ready. For example,
file settings may not have been applied, and we may not yet have a
master node. The readiness service's purpose is to identify when the
node is actually ready for serving requests.

This commit adjusts the end of Node start to wait on the readiness
service being ready. The ramifications are that the main Elasticsearch
thread will not exit until the node is actually ready to serve requests,
and the cli will not exit (when in daemon mode) until the node is ready.

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

ldematte commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -707,6 +711,16 @@ static Settings mergePluginSettings(Map<String, Plugin> pluginMap, Settings orig
+        CountDownLatch ready = new CountDownLatch(1);
+        readinessService.addBoundAddressListener(address -> ready.countDown());
+        try {
+            ready.await();
```
We are not introducing a timeout here - I suppose there are plenty of other places where we already timeout so there is no need for another one here, correct?

rjernst replied:
Yep exactly, there are other timeouts on startup eg systemd has a 75 second default timeout. Right now we will actually report to systemd that the node is ready before the readiness probe responds true, but this change will align them.

ldematte reviewed (approved):
Looks good, only a minor Q

rjernst committed:
Merge branch 'main' into startup/wait_for_ready

rjernst committed:
Merge branch 'main' into startup/wait_for_ready

rjernst committed:
Merge branch 'main' into startup/wait_for_ready

rjernst committed:
start master node first

ldematte reviewed (commented):
Still LGTM, but there is a potentially related test failure you might want to look into

rjernst committed:
Merge remote-tracking branch 'upstream/main' into startup/wait_for_ready

rjernst committed:
Move readiness wait out of Node.start() and into the daemon startup path

Node.start() was blocking indefinitely on the readiness service, which
broke multi-node test clusters where data nodes were started before their
master. The wait belongs in the process startup layer: Elasticsearch.java
now waits for the readiness service between node.start() returning and
sending the SERVER_READY_MARKER to the CLI, so the daemon does not signal
ready until the node is truly ready, without imposing ordering constraints
on Node.start() callers.

rjernst committed:
Add unit tests for waitForNodeReady, make it directly testable

Extract ReadinessService from the Node injector at the call site so
waitForNodeReady takes a ReadinessService directly. This makes it
package-private and independently testable without a real Node.

Two tests verify the causal relationship: one asserts the method blocks
until readiness fires, the other asserts it returns immediately when the
service is already ready.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [9.1](https://github.com/elastic/elasticsearch/pull/151878)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/151878">](https://github.com/elastic/elasticsearch/pull/151878) |
| ✅ |  [9.2](https://github.com/elastic/elasticsearch/pull/151879)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/151879">](https://github.com/elastic/elasticsearch/pull/151879) |
| ✅ |  [8.19](https://github.com/elastic/elasticsearch/pull/151880)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/151880">](https://github.com/elastic/elasticsearch/pull/151880) |
